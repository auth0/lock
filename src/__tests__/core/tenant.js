const getInitTenant = () => require('core/tenant/index').initTenant;

const CLIENT_ID = 'client_id';

const runTest = (initTenant, mockDataFns, client) => {
  initTenant({}, CLIENT_ID, client);
  expect(mockDataFns.initNS.mock.calls.length).toBe(1);
  const tenantInfo = mockDataFns.initNS.mock.calls[0][1].toJS();
  expect(tenantInfo).toMatchSnapshot();
};

describe('initTenant()', () => {
  let initTenant;
  let mockDataFns;
  beforeEach(() => {
    jest.resetModules();

    mockDataFns = {
      initNS: jest.fn(),
      get: jest.fn()
    };
    jest.mock('utils/data_utils', () => ({
      dataFns: () => mockDataFns
    }));

    jest.mock('core/index', () => ({
      findConnection: jest.fn()
    }));
    initTenant = getInitTenant();
  });
  describe('with database connection', () => {
    it('maps connection correctly with defaults', () => {
      const client = {
        connections: {
          database: [
            {
              name: 'test-connection-database',
              strategy: 'auth0'
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
    it('maps connection correctly with all the properties', () => {
      const client = {
        connections: {
          database: [
            {
              allowForgot: false,
              allowSignup: false,
              name: 'test-connection-database',
              requiresUsername: true,
              strategy: 'auth0',
              validation: {
                passwordPolicy: 'test-passwordPolicy',
                username: {
                  min: 4,
                  max: 5
                }
              }
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
    it('fixes validation when values are not numbers', () => {
      const client = {
        connections: {
          database: [
            {
              allowForgot: false,
              allowSignup: false,
              name: 'test-connection-database',
              requiresUsername: true,
              strategy: 'auth0',
              validation: {
                passwordPolicy: 'test-passwordPolicy',
                username: {
                  min: 'foo',
                  max: 'bar'
                }
              }
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
    it('fixes validation when username.min > username.max', () => {
      const client = {
        connections: {
          database: [
            {
              allowForgot: false,
              allowSignup: false,
              name: 'test-connection-database',
              requiresUsername: true,
              strategy: 'auth0',
              validation: {
                passwordPolicy: 'test-passwordPolicy',
                username: {
                  min: 5,
                  max: 4
                }
              }
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
  });
  describe('with enterprise connection', () => {
    it('maps connection correctly', () => {
      const client = {
        connections: {
          enterprise: [
            {
              name: 'test-connection-enterprise',
              domains: 'domains',
              strategy: 'auth0'
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
  });
  describe('with other connection types', () => {
    it('maps connection correctly', () => {
      const client = {
        connections: {
          social: [
            {
              name: 'test-connection-other_type',
              strategy: 'auth0'
            }
          ],
          unknown: [
            {
              name: '??',
              strategy: '??'
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
  });
  describe('with passwordless connection', () => {
    it('maps connection correctly', () => {
      const client = {
        connections: {
          passwordless: [
            {
              name: 'sms',
              strategy: 'sms'
            }
          ]
        }
      };
      runTest(initTenant, mockDataFns, client);
    });
  });
  test('filters clientConnections', () => {
    const client = {
      connections: {
        database: [
          {
            name: 'test-connection-database',
            strategy: 'auth0'
          },
          {
            name: 'test-not-this-one',
            strategy: 'auth0'
          }
        ]
      },
      clientsConnections: {
        [CLIENT_ID]: ['test-connection-database']
      }
    };
    runTest(initTenant, mockDataFns, client);
  });
});
