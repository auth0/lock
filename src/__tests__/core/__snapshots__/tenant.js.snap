// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initTenant() filters clientConnections 1`] = `
Object {
  "connections": Object {
    "database": Array [
      Object {
        "allowForgot": true,
        "allowSignup": true,
        "name": "test-connection-database",
        "passwordPolicy": "none",
        "requireUsername": false,
        "strategy": "auth0",
        "type": "database",
        "validation": null,
      },
    ],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with database connection fixes validation when username.min > username.max 1`] = `
Object {
  "connections": Object {
    "database": Array [
      Object {
        "allowForgot": false,
        "allowSignup": false,
        "name": "test-connection-database",
        "passwordPolicy": "test-passwordPolicy",
        "requireUsername": true,
        "strategy": "auth0",
        "type": "database",
        "validation": Object {
          "passwordPolicy": "test-passwordPolicy",
          "username": Object {
            "max": 15,
            "min": 1,
          },
        },
      },
    ],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with database connection fixes validation when values are not numbers 1`] = `
Object {
  "connections": Object {
    "database": Array [
      Object {
        "allowForgot": false,
        "allowSignup": false,
        "name": "test-connection-database",
        "passwordPolicy": "test-passwordPolicy",
        "requireUsername": true,
        "strategy": "auth0",
        "type": "database",
        "validation": Object {
          "passwordPolicy": "test-passwordPolicy",
          "username": Object {
            "max": 15,
            "min": 1,
          },
        },
      },
    ],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with database connection maps connection correctly with all the properties 1`] = `
Object {
  "connections": Object {
    "database": Array [
      Object {
        "allowForgot": false,
        "allowSignup": false,
        "name": "test-connection-database",
        "passwordPolicy": "test-passwordPolicy",
        "requireUsername": true,
        "strategy": "auth0",
        "type": "database",
        "validation": Object {
          "passwordPolicy": "test-passwordPolicy",
          "username": Object {
            "max": 5,
            "min": 4,
          },
        },
      },
    ],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with database connection maps connection correctly with defaults 1`] = `
Object {
  "connections": Object {
    "database": Array [
      Object {
        "allowForgot": true,
        "allowSignup": true,
        "name": "test-connection-database",
        "passwordPolicy": "none",
        "requireUsername": false,
        "strategy": "auth0",
        "type": "database",
        "validation": null,
      },
    ],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with enterprise connection maps connection correctly 1`] = `
Object {
  "connections": Object {
    "database": Array [],
    "enterprise": Array [
      Object {
        "domains": "domains",
        "name": "test-connection-enterprise",
        "strategy": "auth0",
        "type": "enterprise",
      },
    ],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with other connection types maps connection correctly 1`] = `
Object {
  "connections": Object {
    "database": Array [],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [
      Object {
        "name": "test-connection-other_type",
        "strategy": "auth0",
        "type": "social",
      },
    ],
    "unknown": Array [
      Object {
        "name": "??",
        "strategy": "??",
        "type": "unknown",
      },
    ],
  },
  "defaultDirectory": null,
}
`;

exports[`initTenant() with passwordless connection maps connection correctly 1`] = `
Object {
  "connections": Object {
    "database": Array [],
    "enterprise": Array [],
    "passwordless": Array [],
    "social": Array [],
    "unknown": Array [],
  },
  "defaultDirectory": null,
}
`;
